### ISSUE TITLE
TSP module broken

### ISSUE THREAD

**Issue by yoyo-23** [finance-quote/finance-quote#145]
I found that the Thrift Savings Plan module (TSP.pm, finance::quote v1.49) stopped working sometime in the past month.  After some detective work I determined that the failure was due to retirement of certain funds and the addition of others.  I made modifications to "TSP.pm" and have pasted the updated version below.  The changes were minor: I removed references to the discontinued "L2020" fund, and added "L2025", "L2035", "L2045", "L2055", "L2060", & "L2065".  I use finance-quote mainly through gnucash and the problem with the TSP module caused all my stock/fund updates to fail - there was no clear indication that the problem was with the TSP module and it took some inspired guess work to determine that was the case.  I'm not sure if the unhelpful failure mode is inherent in finance_quote or perhaps in the way it is called from gnucash, but it would be very nice if failure diagnostics could be made a bit more illuminating.  Thanks.

#!/usr/bin/perl -w
#
#    Copyright (C) 1998, Dj Padzensky <djpadz@padz.net>
#    Copyright (C) 1998, 1999 Linas Vepstas <linas@linas.org>
#    Copyright (C) 2000, Yannick LE NY <y-le-ny@ifrance.com>
#    Copyright (C) 2000, Paul Fenwick <pjf@cpan.org>
#    Copyright (C) 2000, Brent Neal <brentn@users.sourceforge.net>
#    Copyright (C) 2001, Rob Sessink <rob_ses@users.sourceforge.net>
#    Copyright (C) 2004, Frank Mori Hess <fmhess@users.sourceforge.net>
#                        Trent Piepho <xyzzy@spekeasy.org>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
#    02111-1307, USA
#
#
# This code is derived from version 0.9 of the AEX.pm module.

require 5.005;

use strict;

package Finance::Quote::TSP;

use vars qw( $TSP_URL $TSP_MAIN_URL %TSP_FUND_COLUMNS %TSP_FUND_NAMES);

use LWP::UserAgent;
use HTTP::Request::Common;
use HTML::TableExtract;

our $VERSION = '1.49'; # VERSION

# URLs of where to obtain information

$TSP_URL = 'https://www.tsp.gov/InvestmentFunds/FundPerformance/index.html';
$TSP_MAIN_URL=("http://www.tsp.gov");

# ENHANCE-ME: The decade target funds like 2020 appear and disappear.
# Better not to hard code them.
#
%TSP_FUND_COLUMNS = (
    L2065   => "L 2065",
    L2060   => "L 2060",
    L2055   => "L 2055",
    L2050   => "L 2050",
    L2045   => "L 2045",
    L2040   => "L 2040",
    L2035   => "L 2035",
    L2030   => "L 2030",
    L2025   => "L 2025",
    LINCOME => "L INCOME",
    G       => "G FUND",
    F       => "F FUND",
    C       => "C FUND",
    S       => "S FUND",
    I       => "I FUND" );

%TSP_FUND_NAMES = (
    L2065    => 'Lifecycle 2065 Fund',
    L2060    => 'Lifecycle 2060 Fund',
    L2055    => 'Lifecycle 2055 Fund',
    L2050    => 'Lifecycle 2050 Fund',
    L2045    => 'Lifecycle 2045 Fund',
    L2040    => 'Lifecycle 2040 Fund',
    L2035    => 'Lifecycle 2035 Fund',
    L2030    => 'Lifecycle 2030 Fund',
    L2025    => 'Lifecycle 2025 Fund',
    LINCOME  => 'Lifecycle Income Fund',
    G        => 'Government Securities Investment Fund',
    F        => 'Fixed Income Index Investment Fund',
    C        => 'Common Stock Index Investment Fund',
    S        => 'Small Capitalization Stock Index Investment Fund',
    I        => 'International Stock Index Investment Fund' );

sub methods { return (tsp => \&tsp) }

{
  my @labels = qw/name date isodate currency close/;
  sub labels { return (tsp => \@labels); }
}

# ==============================================================================
sub tsp {
  my $quoter = shift;
  my @symbols = @_;

  return unless @symbols;

  my(%info, %fundrows);
  my($ua, $reply, $row, $te, $ts, $second_row);

  $ua = $quoter->user_agent;
  $reply = $ua->request(GET $TSP_URL);
  return unless ($reply->is_success);
  $te = HTML::TableExtract->new( headers =>
      ["Date", values %TSP_FUND_COLUMNS] );

  $te->parse($reply->content);

  # First row is newest data, older data follows, maybe there
  # should be some way to get it (in addition to the second_row "close")
  $ts = $te->first_table_found
    || die 'TSP data table not recognised';
  $row = $ts->row(0);
  $second_row = $ts->row(1);

  # Make a hash that maps the order the columns are in
  for(my $i=1; my $key = each %TSP_FUND_COLUMNS ; $i++) {
    $fundrows{$key} = $i;
  }
  
  foreach (@symbols) {
    my $symbol = uc $_;

    if(exists $fundrows{$symbol}) {
      $info{$_, 'success'} = 1;
      $info{$_, 'name'} = $TSP_FUND_NAMES{$symbol};
      $quoter->store_date(\%info, $_, {usdate => $$row[0]});
      ($info{$_, 'last'} = $second_row->[$fundrows{$symbol}]) =~ s/[^0-9]*([0-9.,]+).*/$1/s;
      $info{$_, 'currency'} = 'USD';
      $info{$_, 'method'} = 'tsp';
      $info{$_, 'source'} = $TSP_MAIN_URL;
      $info{$_, 'symbol'} = $_;
    } 
    else {
      $info{$_, 'success'} = 0;
      $info{$_, 'errormsg'} = "Fund name unknown";
    }
  }
  return %info if wantarray;
  return \%info;
}
1;

=head1 NAME

Finance::Quote::TSP Obtain fund prices for US Federal Government Thrift Savings Plan

=head1 SYNOPSIS

    use Finance::Quote;

    $q = Finance::Quote->new;

    %info = Finance::Quote->fetch("tsp","c"); #get value of C "Common Stock Index Investment" Fund

=head1 DESCRIPTION

This module fetches fund information from the "Thrift Savings Plan"

    http://www.tsp.gov

using its fund prices page

    https://www.tsp.gov/investmentfunds/shareprice/sharePriceHistory.shtml

The quote symbols are

    C          common stock fund
    F          fixed income fund
    G          government securities fund
    I          international stock fund
    S          small cap stock fund
    L2025      lifecycle fund year 2025
    L2030      lifecycle fund year 2030
    L2035      lifecycle fund year 2035
    L2040      lifecycle fund year 2040
    L2045      lifecycle fund year 2045
    L2050      lifecycle fund year 2050
    L2055      lifecycle fund year 2055
    L2060      lifecycle fund year 2060
    L2065      lifecycle fund year 2065
    LINCOME    lifecycle income fund

=head1 LABELS RETURNED

The following labels may be returned by Finance::Quote::TSP :

    name        eg. "Lifecycle 2050 Fund"
    date        latest date, eg. "21/02/10"
    isodate     latest date, eg. "2010-02-21"
    last        latest available price, eg. "16.1053"
    currency    "USD"
    method      "tsp"
    source      TSP URL

=head1 SEE ALSO

Thrift Savings Plan, http://www.tsp.gov

=cut

**Comment by jhaiduce** [comment]
I just submitted your changes in a pull request (https://github.com/finance-quote/finance-quote/pull/146).

I think the unhelpful failure mode is specific to GnuCash. If you run gnc-fq-dump on a TSP fund you get a more helpful error message:

```
$ gnc-fq-dump tsp TSPL2040
TSP data table not recognised at /home/jhaiduce/perl5/lib/perl5/Finance/Quote/TSP.pm line 108.
```

This suggests to me that Finance::Quote is raising a useful exception and GnuCash is throwing out the detailed information and just saying that something failed.

**Comment by jhaiduce** [comment]
@yoyo-23, when I implemented the changes you described, it allows gnucash to import all the other quotes, but it says "Unable to retrieve quotes for these items" and then lists all my TSP funds. 

gnc-fq-dump is able to retrieve some data but not the actual prices:
`$ gnc-fq-dump tsp TSPIFUND
Finance::Quote fields Gnucash uses:
    symbol: TSPIFUND (deduced)   <=== required
      date: ** missing **        <=== recommended
  currency: ** missing **        <=== required
Use of uninitialized value $last in printf at /usr/bin/gnc-fq-dump line 120.
      last:                      <=\       
Use of uninitialized value $nav in printf at /usr/bin/gnc-fq-dump line 121.
       nav:                      <=== one of these
Use of uninitialized value $price in printf at /usr/bin/gnc-fq-dump line 122.
     price:                      <=/        
  timezone:                      <=== optional

** This stock quote cannot be used by GnuCash!`

Did you get the same result?

**Comment by jhaiduce** [comment]
> @yoyo-23, when I implemented the changes you described, it allows gnucash to import all the other quotes, but it says "Unable to retrieve quotes for these items" and then lists all my TSP funds.
> 
> gnc-fq-dump is able to retrieve some data but not the actual prices:
> 
> ```
> $ gnc-fq-dump tsp TSPIFUND
> Finance::Quote fields Gnucash uses:
>     symbol: TSPIFUND (deduced)   <=== required
>       date: ** missing **        <=== recommended
>   currency: ** missing **        <=== required
> Use of uninitialized value $last in printf at /usr/bin/gnc-fq-dump line 120.
>       last:                      <=\       
> Use of uninitialized value $nav in printf at /usr/bin/gnc-fq-dump line 121.
>        nav:                      <=== one of these
> Use of uninitialized value $price in printf at /usr/bin/gnc-fq-dump line 122.
>      price:                      <=/        
>   timezone:                      <=== optional
> 
> ** This stock quote cannot be used by GnuCash!
> ```

I see now what the problem was. The fund symbols in Finance::Quote::TSP were changed in commit a7ca16ec183d2e8efb3f752e041539c16272d8a1. So if I change 'TSPIFUND' to just 'I' I get a quote:

```
$ gnc-fq-dump tsp I
Finance::Quote fields Gnucash uses:
    symbol: i                    <=== required
      date: 07/07/2020           <=== recommended
  currency: USD                  <=== required
      last: 30.0523              <=\       
       nav:                      <=== one of these
     price:                      <=/        
  timezone:                      <=== optional
[jhaiduce@picard perl5]$ gnc-fq-dump tsp i
Finance::Quote fields Gnucash uses:
    symbol: i                    <=== required
      date: 07/07/2020           <=== recommended
  currency: USD                  <=== required
      last: 30.0523              <=\       
       nav:                      <=== one of these
     price:                      <=/        
  timezone:                      <=== optional
```

After I update the symbols in the Gnucash security editor then fetching quotes works from inside Gnucash too.

**Comment by yoyo-23** [comment]
It sounds like you've figured it out, but FYI, no I didn't have a problem downloading quotes for any of my TSP funds (I,F,C,G,S) once I made the changes to TSP.pm - works like a charm.  David

    On Wednesday, July 8, 2020, 3:49:21 AM UTC, John Haiducek <notifications@github.com> wrote:  
 
 



@yoyo-23, when I implemented the changes you described, it allows gnucash to import all the other quotes, but it says "Unable to retrieve quotes for these items" and then lists all my TSP funds.

gnc-fq-dump is able to retrieve some data but not the actual prices:
$ gnc-fq-dump tsp TSPIFUND
Finance::Quote fields Gnucash uses:
    symbol: TSPIFUND (deduced)   <=== required
      date: ** missing **        <=== recommended
  currency: ** missing **        <=== required
Use of uninitialized value $last in printf at /usr/bin/gnc-fq-dump line 120.
      last:                      <=\       
Use of uninitialized value $nav in printf at /usr/bin/gnc-fq-dump line 121.
       nav:                      <=== one of these
Use of uninitialized value $price in printf at /usr/bin/gnc-fq-dump line 122.
     price:                      <=/        
  timezone:                      <=== optional

** This stock quote cannot be used by GnuCash!


I see now what the problem was. The fund symbols in Finance::Quote::TSP were changed in commit a7ca16e. So if I change 'TSPIFUND' to just 'I' I get a quote:
$ gnc-fq-dump tsp I
Finance::Quote fields Gnucash uses:
    symbol: i                    <=== required
      date: 07/07/2020           <=== recommended
  currency: USD                  <=== required
      last: 30.0523              <=\       
       nav:                      <=== one of these
     price:                      <=/        
  timezone:                      <=== optional
[jhaiduce@picard perl5]$ gnc-fq-dump tsp i
Finance::Quote fields Gnucash uses:
    symbol: i                    <=== required
      date: 07/07/2020           <=== recommended
  currency: USD                  <=== required
      last: 30.0523              <=\       
       nav:                      <=== one of these
     price:                      <=/        
  timezone:                      <=== optional

After I update the symbols in the Gnucash security editor then fetching quotes works from inside Gnucash too.
